### PR TITLE
Retriable as a class extension

### PR DESCRIPTION
**Pull request by ridiculous**
Extend the Retriable DSL with a class extension

```ruby
class Example
  extend Retriable
  def perform
    # some work
  end
  retriable :perform, on: Net::OpenTimeout
end
```

### CODE REVIEW

File: lib/retriable/ext.rb [hunk 1]
```diff
@@ -0,0 +1,18 @@
+module Retriable
+  module Ext
```
[the remark this answers is not in the document]
Comment (apurvis) [reply lib/retriable/ext.rb]: there's no better naming than `Ext`?  i think that is kind of an unacceptably vague name.  `Ext` tells me nothing about what this module does.

**Comment by ridiculous** [comment]
Sure, I can fix em. Do you like the idea tho?

**Comment by apurvis** [comment]
i like the idea generally and have thought of doing exactly this before, actually, so nice work.

File: lib/retriable/ext.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (ridiculous) [reply lib/retriable/ext.rb]: I'd prefer if we could `extend Retriable` directly. It'd require changing scope of module methods, so not everything ends up extended on the host object. The class method `retriable` that get's extended is different than the module function `Retriable.retriable` that is the public API of the gem. I'm slightly concerned about breaking peoples usage of it, if for instance they have been mixing it into their classes already

**Comment by ridiculous** [comment]
@apurvis consolidated the updates into `Retriable`. Let me know what you think

**Comment by kamui** [comment]
@ridiculous Thanks for the PR and sorry for the late reply. I can see this being useful, the only request I have is to make the extension a separate module/file like the kernel extension.

**Comment by apurvis** [comment]
also needs documentation in the README

**Comment by ridiculous** [comment]
You don't think it's simpler to keep it as one? From a user's perspective, it's nice to only have to learn one module. You can call it directly via `Retriable.retriable` or `extend Retriable`. I think that's pretty intuitive. Are you worried it complicates the implementation?

**Comment by kamui** [comment]
I think the usual case is that most people wont use this extension, so it doesn't make sense to bring in a code path that wont be used or executed. If you want the extension, then you bring it in and only then does it come in.

I think it's a cleaner separation and it prevents any possible code paths that support this feature from causing any issues now or in the future if new features or refactoring is made on top of it.

**Comment by ridiculous** [comment]
That's unfortunate, cause I much prefer the extension approach. It allows you add retries without altering a method definition. I think if people had the option (with this gem) then it'd become the preferred approach. It makes it so easy to add retries, even to dynamically generated methods. The [memoist](https://github.com/matthewrudy/memoist) does this to great success. It's a clean and easy to use interface.

**Comment by kamui** [comment]
Maybe you're not understanding my ask. I'm just saying make this possible:

```ruby
require "retriable/core_ext/class_extension"

class Example
  extend Retriable::ClassExtension
  def perform
    # some work
  end
  retriable :perform, on: Net::OpenTimeout
end
```

Just move out the `retriable` method with the `define_method` calls into `core_ext/class_extension.rb` like how the `core_ext/kernel.rb` works. I'm not saying this can't be in this gem, just don't put the class extending code into the main `Retriable` module.

**Comment by ridiculous** [comment]
I understand. I'm just saying that's clunky. Retriable is a module, why can't I extend it directly?

**Comment by kamui** [comment]
I think because it's not currently the standard use case for this gem, using it in a class would be an extended use case. Adding this ability adds a code path to the base module that may be unnecessary for a majority of use cases. I also can't anticipate if this will have any bugs or issues in different context, so it seems safer to start as an extension you need to require see how it goes. If it gets adopted as a standard use case and if there are no incident reports, we can revisit making it supported in the base module.

**Comment by ridiculous** [comment]
The problem is: once it's out there, it's a pain to take away. If we end up bringing it into core, it'd make upgrading the gem harder for users. Thinking long term, at least for me, ideally it'd always be in the core module. We'd maintain control over how the gem's structured internally. And by control I mean ability to change things without breaking end-users.

Say we want to support a robust module interface, where the gem can be used in any form you may normally use a module:
```ruby
Retriable.retriable # module function (current)
class Example
  extend Retriable  # class extension
  include Retriable # mixin
end
```
That seems like a pretty good interface. Flexible and intuitive. We can do this if we bring everything into core. 

Whether it's intended or not, the gem can currently be used as a mixin. Some users may already be doing that.

We could restructure core to support the example above, with something like this:
```ruby
module Retriable
  module BaseMethods
    def retriable
      'Base method'
    end
  end

  module Ext
    def retriable
      'class extension'
    end
  end

  extend BaseMethods

  def self.included(base)
    base.send :include, Retriable::BaseMethods
  end

  def self.extended(base)
    base.send :extend, Retriable::Ext
  end
end
```
I think we should decide sooner, rather than later, if we wanna support this as a feature of the gem. And if we do, then let's do it right.

**Comment by ridiculous** [comment]
Any update on this? Yay or nay?

**Comment by apurvis** [comment]
FWIW I tend to agree with @ridiculous that this is a pretty clean way to do it...  and I also remember that we had a similar discussion about adding features as extensions over on #39 .  Eventually, after some substantial time for consideration, @kamui agreed that it was cleaner to just add `with_context()` the main gem and not require a separate `include`.  wondering if this conversation might end up the same way, though I do think this is more complicated case given that now there would be both a class and module instance method on the main module.

**Comment by apurvis** [comment]
maybe a good way to avoid confusion would be to name the module instance method something else... perhaps something along the lines of `make_method_retriable()` or something.

**Comment by apurvis** [comment]
also FWIW if this feature existed in a way that also included the context feature - such that i could call `make_method_retriable_with_context(method_name, :context)` - i think it could potentially see quite a lot of use.  i definitely would use it all the time.

**Comment by ridiculous** [comment]
Thanks for following up on this! I'd love to get this moving again. Some questions:
> maybe a good way to avoid confusion would be to name the module instance method something else... perhaps something along the lines of make_method_retriable() or something.

I actually think this makes it more confusing, because there'd be more methods, one on class and another on instance. We'd need to remember more names, and be more context aware ourselves. I'd rather let the gem do it, since it can. 

And learning one interface is just easier.
> also FWIW if this feature existed in a way that also included the context feature - such that i could call make_method_retriable_with_context(method_name, :context) - i think it could potentially see quite a lot of use. i definitely would use it all the time.
I'm not sure what you mean by this...? What is `:context`, a proc?

**Comment by ridiculous** [comment]
Got it, thanks. Didn't realize this gem has contexts. Very nice! I'm using the same thing in my version, except there's still just one interface, that accepts classes or symbols, and maps the symbols to contexts.

> yes there would be more methods, but they would do different things (one would retry a block; one would retry a method).

Are they really that different? Calling a method if no blocks given seems like a reasonable default. Think of `Enumerable`, where `reduce` takes a block or optional second argument. People are familiar with the concept, I think.

**Comment by ridiculous** [comment]
If we think of this as tool that appears to be baked into the language, it'll be a huge success!

**Pull request closed without merging**